lib/pengine/unpack.c

```diff
@@ -4971,9 +4988,23 @@ add_node_attrs(const xmlNode *xml_obj, pcmk_node_t *node, bool overwrite,
                          cluster_name);
     }
 
-    pe__unpack_dataset_nvpairs(xml_obj, PCMK_XE_INSTANCE_ATTRIBUTES, &rule_data,
-                               node->priv->attrs, NULL, overwrite,
-                               scheduler);
+    if (overwrite) {
+        /* @TODO Try to reorder some unpacking so that we don't need the
+         * overwrite argument or to unpack into a temporary table
+         */
```

Comment:
Instead of unpacking permanent attributes when unpacking node configuration and then unpacking transient attributes when unpacking node state, it should be possible to wait until the end of unpacking node state, then for each node, unpack transient attributes then permanent attributes

Reply:
That was my thought but I didn't feel like tracing through the EAGAIN logic of `unpack_node_history()` which is what calls `unpack_handle_remote_attrs()`

Reply:
I'm thinking it would be done after the last unpack_node_history() call (i.e. separate node attribute unpacking from config/history unpacking). I'd have to double-check no node attributes are used before that point, but I expect it would be problematic if they were.

Reply:
> used before that point

At least somewhat. `unpack_handle_remote_attrs()` (via `unpack_node_history()`) and `unpack_transient_attributes()` (via `unpack_node_state()`) modify the standby, maintenance, and (for remote) `pcmk__node_probes_allowed` booleans/flags.

The standby flag is used for logging purposes in `determine_online_status()`. The maintenance boolean is used for that purpose, and it also affects `process_rsc_state()` (via `unpack_lrm_resource()`).

I'm sure this can be done but it's not quite as straightforward as one would hope.

Reply:
More specifically into a new function like: for each node, call add_node_attrs() for permanent attributes, call add_node_attrs() for transient attributes, do the processing stuff

Reply:
The logging of node status currently in determine_online_status() would also move to that same new function (ideally functionized itself)

But all that can wait until we get to best practices for the unpacking code, which will be a massive project